### PR TITLE
[ie/facebook] improve format sorting

### PR DESCRIPTION
**Pull request by fireattack**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Improve how the formats are sorted.

This IE extracts video formats from:

1. plain URLs which are denoted as "HD" (usually 720p) and SD (anything lower), and they're assigned quality = 0 and 1 respectively. No other technical info can be easily extracted.
2. formats extracted from MPD, which have all the info including resolution etc. It can have up to 1080p (potentially higher?) resolution. They have no "quality" assigned at the moment.

The IE currently also have a sort order of `res, quality`, which makes sure that all the formats from MPD will have higher priority than the "sd" and "hd" formats (because they have no `res` key). This is desired because (best) MPD ones have better, or at least equal, quality of video than SD/HD ones.

This works fine under default settings. However, this causes a problem: if the user have assigned `-S` for other reasons, even if it is similar to what the default has (`lang,quality,res,fps,...`), it will overwrite the extractor's and therefore causes these "sd" and "hd" to have higher priority than MPD formats. Because the default sort order have quality before res.

This PR changes it to:
1. Assign quality = 2 to all the MPD formats. SD and HD formats still have quality of 0 and 1, respectively. 
2. Get rid of `_format_sort_fields`.

This makes the default behavior identical to previous: MPD formats will still be ahead of HD/SD due to have better "quality" now, instead of due to having "res". But no longer cause any issues when the user have any sensible custom `-S`.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
📹🛠️📈

Fix Facebook GraphQl video extraction by adding default quality to formats. Update `yt_dlp/extractor/facebook.py` to handle formats without quality field.

> _`quality` added_
> _Facebook videos fixed now_
> _Autumn of errors_

### Walkthrough
* Add a default quality value to formats without quality field ([link](https://github.com/yt-dlp/yt-dlp/pull/8074/files?diff=unified&w=0#diff-ca6052d2c84eb4ac45aaf04463a4eaecdc75fb3ae4766e234917d8ab131b60ecL507-R509)) in `facebook.py` to avoid sorting errors

</details>

### CODE REVIEW

**Comment by bashonly** [comment]
What if other people's use cases depend on the previous `quality` distribution / extractor sort order? e.g. if there are people regularly using `-S quality` to download the `hd` format

**Comment by fireattack** [comment]
That's definitely a valid concern if someone really wants that instead of best from MPD; I assume they can use `-S quality:1`?

**Comment by bashonly** [comment]
format `quality` is also set in this block of code starting at line 719:
```py
            for quality in ('sd', 'hd'):
                for src_type in ('src', 'src_no_ratelimit'):
                    src = f[0].get('%s_%s' % (quality, src_type))
                    if src:
                        preference = -10 if format_id == 'progressive' else -1
                        if quality == 'hd':
                            preference += 5
                        formats.append({
                            'format_id': '%s_%s_%s' % (format_id, quality, src_type),
                            'url': src,
                            'quality': preference,
                            'height': 720 if quality == 'hd' else None
                        })
```
won't this patch change how these formats are sorted by default?

**Comment by fireattack** [comment]
Yeah, I had a glance at the part of the code.

Unfortunately I cannot find any tests that will trigger that part (not even https://github.com/yt-dlp/yt-dlp/issues/795 which was directly referring to this, now the video in the ticket only have SD quality)

But by the look at it, it will generate -10 (sd+progressive), -5 (hd+progressive), -1 (sd+other format), and 4 (hd+other format) respectively. 

I have no idea what "other format" could be, and why progressive one should be penalized to -10.
It was added in https://github.com/ytdl-org/youtube-dl/commit/44d6dd08b299ccf17eb04901cf09a8d333769783 to fix https://github.com/ytdl-org/youtube-dl/issues/7252 but no reason is given about it.

And the +5 part was introduced at this commit https://github.com/yt-dlp/yt-dlp/commit/c24883a1c0e650a41b54e7cdec8427b046c27bc1: I don't think too much thought was put into the choice of number, which explains why it could end up as "4" (or "5" at the time).

"0" for other formats was changed to -1 in https://github.com/yt-dlp/yt-dlp/commit/8a2d992389c37f5f99f5c74677f7900f1ae45f94 (together with the sorting of `res, quality`. ) Again, I don't think the choice of the numbers are deliberate. 

I think this part of code probably would never been reached today, due to changes on facebook. But I will just leave them here, and modify the value of them to:

-10 (sd+progressive), -9 (hd+progressive), -1 (sd+other format), and 0 (hd+other format).

This way, the latter two will be align with SD/HD acquired through other means, and progressive kept penalized for whatever reason it was. -9 and -5 makes no difference anyway, and simplifies the code.

What do you think?

**Comment by Grub4K** [comment]
I agree that we should remove the custom sort order.

However, we can remove the `if 'quality' not in f:` branch by setting the quality of all other formats below the default of `-1`. This makes it easier to understand the header comment, allows new formats to be sorted properly and keeps the quality code in the same place.
```diff
diff --git a/yt_dlp/extractor/facebook.py b/yt_dlp/extractor/facebook.py
index e84b78fa5..c98a7c2c0 100644
--- a/yt_dlp/extractor/facebook.py
+++ b/yt_dlp/extractor/facebook.py
@@ -504,8 +504,6 @@ def process_formats(info):
             # Downloads with browser's User-Agent are rate limited. Working around
             # with non-browser User-Agent.
             for f in info['formats']:
-                if 'quality' not in f:
-                    f['quality'] = 2
                 f.setdefault('http_headers', {})['User-Agent'] = 'facebookexternalhit/1.1'
 
         def extract_relay_data(_filter):
@@ -553,7 +551,7 @@ def parse_graphql_video(video):
                         else:
                             formats.append({
                                 'format_id': format_id,
-                                'quality': q(format_id),
+                                'quality': q(format_id) - 3,
                                 'url': playable_url,
                             })
                     extract_dash_manifest(video, formats)
@@ -720,7 +718,7 @@ def parse_attachment(attachment, key='media'):
                 for src_type in ('src', 'src_no_ratelimit'):
                     src = f[0].get('%s_%s' % (quality, src_type))
                     if src:
-                        preference = -10 if format_id == 'progressive' else 0
+                        preference = -10 if format_id == 'progressive' else -3
                         if quality == 'hd':
                             preference += 1
                         formats.append({
```
Perhaps we could even add additional comments explaining why these qualities were chosen, giving reference to the unknown `progressive` downgrade.

**Comment by fireattack** [comment]
Thanks for the suggestion, @Grub4K ! I've incorporated these changes together with a comment. 

BTW am I supposed to assign a reviewer myself? Or you guys will take care of it.

**Comment by bashonly** [comment]
We'll handle it

File: yt_dlp/extractor/facebook.py [hunk 1]
```diff
@@ -552,7 +551,9 @@ def parse_graphql_video(video):
                         else:
                             formats.append({
                                 'format_id': format_id,
-                                'quality': q(format_id),
+                                # downgrade quality of old sd, hd formats to be lower than formats from DASH
+                                # since they are only up to 720p and no tech info can be extracted.
+                                'quality': q(format_id) - 3,
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:556]:
```suggestion
                                # sd, hd formats w/o resolution info should be deprioritized below DASH
                                'quality': q(format_id) - 3,
```

File: yt_dlp/extractor/facebook.py [hunk 2]
```diff
@@ -719,9 +720,11 @@ def parse_attachment(attachment, key='media'):
                 for src_type in ('src', 'src_no_ratelimit'):
                     src = f[0].get('%s_%s' % (quality, src_type))
                     if src:
-                        preference = -10 if format_id == 'progressive' else -1
+                        # downgrade quality of old sd, hd formats to be lower than formats from DASH
+                        # since they are only up to 720p and no tech info can be extracted.
+                        preference = -10 if format_id == 'progressive' else -3
```
Comment (bashonly) [inline yt_dlp/extractor/facebook.py:725]:
```suggestion
                        # sd, hd formats w/o resolution info should be deprioritized below DASH
                        # TODO: investigate if progressive or src formats still exist
                        preference = -10 if format_id == 'progressive' else -3
```

**Review by bashonly — APPROVED** [review]

**Review by Grub4K — APPROVED** [review]

**Pull request merged**